### ISSUE TITLE
ENH: Check GeoSeries.intersects(GeoSeries) against all geometries, not only pair-wise

### ISSUE THREAD

**Issue by jiezhengxu** [geopandas/geopandas#972]
edit (20/03/04) by @martinfleis: _See the discussion below for the context._
<hr>

```
def merge_boundry(image_path, vector_path, clip_size):
    path_img=os.path.dirname(image_path)
    L=os.listdir(image_path)
    for i in L:  
        dname=os.path.join(path_img,i)
        #add
        with rasterio.open(dname) as src:
            transform = src.transform
            crs = src.crs
            img_arr = reshape_as_image(src.read())
            out_transform= _clip_transform(src.meta['transform'], clip_size)
            origin_height, origin_width, _ =img_arr.shape
        
            origin_height=origin_height-2*clip_size
            origin_width=origin_width-2*clip_size
        
            min_x=out_transform[2]
            min_y=out_transform[5]
            max_x=min_x+out_transform[0]*origin_height
            max_y=min_y+out_transform[4]*origin_width
        
           #every location coordinate
            left_top=min_x, min_y
            left_bottom=min_x, max_y 
            right_top=max_x, min_y
            right_bottom=max_x, max_y    
        
            p1, p2, p3, p4, p5= left_top, left_bottom, right_bottom, right_top, left_top
            draw_frame=LineString([p1,p2,p3,p4,p5])
            #invert format
            draw_frame=geopandas.GeoSeries(draw_frame)
            draw_frame.crs=crs
    
    path_vector=os.path.dirname(vector_path)
    vector_name=os.listdir(vector_path)
    
    for vector in vector_name:
        draw_frame_buffer=draw_frame.buffer(1)
        all_path_merge=os.path.join(path_vector, vector)
#         vector_data=geopandas.read_file(all_path_merge)
        vector_data=geopandas.GeoDataFrame.from_file(all_path_merge)
        block_intersect=draw_frame_buffer.intersects(vector_data)
        print(block_intersect)
```
Result:
0      True
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
dtype: bool
0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
dtype: bool
0    False
dtype: bool
0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
dtype: bool
0      True
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
dtype: bool

reason :
 why not true, only return false ,in fact, exits true

**Comment by jorisvandenbossche** [comment]
@jiezhengxu without more context it is very difficult to understand your issue or question. Can you explain what you are trying to do? And can you also try to provide a reproducible example? (some code that we can run)

**Comment by b4l** [comment]
@jorisvandenbossche I think i met the same issue when intersecting two GeoDataFrames after reprojecting using Pandas 1.0.1 and GeoPandas 0.7. Here is an example:

`import pandas as pd
import geopandas as gpd
import requests
import io
import zipfile
import tempfile

# some pois
pois = pd.DataFrame(
    {
        "latitude": [47.58, 46.23, 47.46, 46.91, 47.44],
        "longitude": [7.52, 6.10, 8.54, 7.49, 8.23],
    }
)
pois = gpd.GeoDataFrame(
    pois, geometry=gpd.points_from_xy(pois.longitude, pois.latitude)
)
pois.crs = "EPSG:4326"

# some polygons
url = "https://data.geo.admin.ch/ch.swisstopo.swissboundaries3d-land-flaeche.fill/shp/2056/ch.swisstopo.swissboundaries3d-land-flaeche.fill.zip"
r = requests.get(url)
z = zipfile.ZipFile(io.BytesIO(r.content))
tmp = tempfile.mkdtemp()
z.extractall(tmp)
ch = gpd.read_file(
    tmp + "/SHAPEFILE_LV95_LN02/swissBOUNDARIES3D_1_3_TLM_LANDESGEBIET.shp"
)
ch.geometry = ch.buffer(10000).to_crs("EPSG:4326")

# plot
base = ch.plot(color='white', edgecolor='black')
pois.plot(ax=base, marker='o', color='red', markersize=5)

# intersection
pois.intersects(ch)

# what i expect
pois.apply(lambda x: ch.intersects(x.geometry).any(), axis="columns")`

**Comment by martinfleis** [comment]
@b4l `gdf1.intersects(gdf2)` does pair-wise intersection. Rows are matched and only geometries within the same rows are checked against each other. But I agree that it is not what one might expect. The solution might be to add a new keyword to have `gdf.intersects(gdf2, any=True)` which would do what you expected (and could even use spatial index under the hood to do so).

**Comment by b4l** [comment]
Thank you for the insights @martinfleis. I think the [spatial predicates](https://en.wikipedia.org/wiki/DE-9IM#Spatial_predicates) should be handled by default as set operations, whereas the elementwise variant could be accessed through an optional keyword. What do you think?

**Comment by martinfleis** [comment]
I feel the same about it, but to changing this behaviour to default on `.any()` would be backward incompatible and might break a lot of code if implemented. Moreover, this element wise logic is something we inherit from pandas and having some functions working like that and other not is something I'd like to avoid.

**Comment by b4l** [comment]
As far as I know, there are many functions in pandas that do not operate elementwise, like all the merge, join, append etc. The 1.0 Milestone would probably be a good opportunity to introduce these changes. It might be painfull, though even more justifying a counterintuitive api to the user.

**Comment by ljwolf** [comment]
For a bit of context, the align-then-relate pattern used in geopandas is a common rough edge I think in part because it's not how [PostGIS works](https://postgis.net/docs/ST_Touches.html) for spatial predicates (in addition to being different from the DE9IM semantics). This is why it's all-pairs by default in R's `sf`, too.

I understand how we got here, but I can see how it's a confusing behavior for users, so an all-pairs relates between geoseries by default would definitely make sense for a 1.0 release, but it's a big ask... like @martinfleis, am concerned that this is a huge breakage.
 
If so, a keyword argument is definitely preferred (though I wouldn't want `any()`, I'd want `pairwise=True`. Users should explicitly use `df.intersects(df2, pairwise=True).any(axis=1)` for
> Does `df` intersect __any element__ of `df2`?

Alternatively, a more explicit separation could put fast index-enabled pairwise predicates under a specific namespace in the dataframe, like `df.pairwise.intersects(df2)`. This is used elsewhere [say, sklearn](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics.pairwise) to separate the functions that might apply to a single coordinate pair versus applying the function to all pairs between two matrices of coordinates.

**Comment by b4l** [comment]
I would prefer `df.intersects(df2, pairwise=True)` with the default value set to `False` for the set operation. Alternatively the default could be set to `True` and changed later to avoid breakage and time the transition.

**Comment by martinfleis** [comment]
I think we could do `df.intersects(df2, pairwise=False)` with a deprecation warning that default will change. I would keep it there for a few releases and then change it. It may be 1.0.0 or 0.11.0 or whatever will fit to give enough time for a transition.

**Comment by ljwolf** [comment]
If we use `pairwise=True` to mean row-by-row, that's different from the semantics of `sklearn.metrics.pairwise` as well as the implied meaning of `scipy.spatial.distance.pdist`. I don't think that's a good idea... What about `all_pairs=True`?

**Comment by b4l** [comment]
What about `paired=True` relatet to `paired_distance` in sklearn? scipy has cdist which seems odd.

**Comment by martinfleis** [comment]
Good points. I like `paired` as it is a bit more explanatory than `all_pairs`.

**Comment by b4l** [comment]
Perhaps it might be useful to take other comparison methods into account as well in designing the interface, like for example bbox intersection.  The latter could be achieved by another parameter or combined in a string typed mode parameter for example.

**Comment by martinfleis** [comment]
I have recently played with this and came up with a possible implementation. See https://gist.github.com/martinfleis/abc7cdbf9f9266bf9ed369080eec7cea 

I am not sure about the API yet, so there are two main questions to resolve (from an implementation point of view):

1. Which options should we offer? Existing + indices or existing + indices + "any"?
2. How should keywords look like?

Note that there might be some temporary workarounds as pygeos does not support all predicates yet (https://github.com/pygeos/pygeos/pull/157).

And then there's the question of change of default...

**Comment by brendan-ward** [comment]
Follow up note: the additional predicates that make sense in the context of a spatial index are now available as of [pygeos version 0.8](https://github.com/pygeos/pygeos/blob/master/CHANGELOG.rst#version-08-2020-09-06).

**Comment by felix-ht** [comment]
any updates on this?

**Comment by martinfleis** [comment]
@felix-ht There is a draft PR and a related discussion in https://github.com/geopandas/geopandas/pull/1674 I need to update whenever I have more time for that. If you need to use it now, you can get the code from that PR.

**Comment by SlowMo24** [comment]
To my understanding this issue discussed the design choice of doing row-wise geometry operations vs set-wise. And while most other frameworks do set-wise, Geopandas, for historical reasons, does row wise.

As a work-around one can use the `GeoDataFrame.sjoin` which will do a set-wise comparison.

But I'm a bit confused why #1674 solved the issue at hand because it finally implemented a change to the sindex-query which will, as far as I understand, not give exact results but just based on BBOX (as any index should). So its only one step on the way of set-wise comparison.

**Comment by martinfleis** [comment]
> just based on BBOX

no, you can specify `predicate` keyword which does predicate-based check providing set-wise comparison.
